sympy/ntheory/residue_ntheory.py

```diff
@@ -1392,3 +1414,103 @@ def quadratic_congruence(a, b, c, p):
         for j in root:
             res.add(j % p)
     return sorted(res)
+
+
+def _general_congruence_prime(coefficients, p):
+    """A helper function used by general_congruence.
+    It returns the root of a polynomial modulo prime number
+    by naive search from [0, p).
+
+    Parameters
+    ==========
+    coefficients : list of integers
+    p : prime number
+    """
+
+    roots = []
+    rank = len(coefficients)
+    for i in range(0, p):
+        f_val = 0
+        for coeff in range(0,rank - 1):
+            f_val = (f_val + pow(i, int(rank - coeff - 1), p) * coefficients[coeff]) % p
+        f_val = f_val + coefficients[-1]
+        if f_val % p == 0:
+            roots.append(i)
+    return roots
+
+
+def _diff_poly(root, coefficients, p):
+    """A helper function used by general_congruence.
+    It returns derivative of the polynomial at root, then applying modulo p.
+
+    Parameters
+    ==========
+    coefficients : list of integers
+    p : prime number
+    root : integer
+    """
+
+    diff = 0
+    rank = len(coefficients)
+    for coeff in range(0, rank - 1):
+        if not coefficients[coeff]:
+            continue
+        diff = (diff + pow(root, rank - coeff - 2, p) * (rank - coeff - 1) * coefficients[coeff]) % p
+    return diff % p
+
+
+def _val_poly(root, coefficients, p):
+    """A helper function used by general_congruence.
+    It returns value of the polynomial at root, then applying modulo p.
+
+    Parameters
+    ==========
+    coefficients : list of integers
+    p : prime number
+    root : integer
+    """
+    rank = len(coefficients)
+    f_val = 0
+    for coeff in range(0, rank - 1):
+        f_val = (f_val + pow(root, rank - coeff - 1, p) * coefficients[coeff]) % p
+    f_val = f_val + coefficients[-1]
+    return f_val % p
+
+
+def general_congruence(expr, m):
+    """
+    Find the solutions to a polynomial congruence equation modulo m.
+    Parameters
```

Comment:
does the guide suggest a blank line before "Parameters"?

Reply:
yes, you are right. I have corrected it now.

Reply:
I think there should be a blank line.

The function signature here seems inconsistent with other functions in this module and elsewhere. Generally a function that guesses the symbol from an equation will also have a symbol argument that can be used to specify the symbol e.g. `solve(eq, x)`. The symbol can be optional if there is only one free symbol in the equation.

Most functions here in this module don't seem to take an expression though but rather take arguments that are integers. The equivalent here would be a list of integers representing the polynomial coefficients.

Also would `polynomial_congruence` be a more appropriate name than `general_congruence`?

Reply:
yes polynomial_congruence will be more appropriate, i will change it.

Reply:
in this function taking expression as argument can be helpful. It allows user to give
 `(1+ x**2 +x**3)**25` as an input. If we change the input to integers the user will have to calculate the coefficients, which might be slightly inconvenient.

Reply:
You can use `Poly((1+ x**2 +x**3)**25).all_coeffs()` for that. The question is what kind of API this module makes and what is it used for. A higher-level function could go somewhere else that takes expressions and converts them e.g.
```julia
In [45]: solveset(x**2 % 3 - 1, x, Integers)                                                             
Out[45]: {3⋅n + 1 | n ∊ ℤ} ∪ {3⋅n + 2 | n ∊ ℤ}
```

Reply:
> > Poly((1+ x**2 +x**3)**25).all_coeffs()
> 
> I'm not sure if it applies here, but maybe one could use the unexpanded form. (Ignore if that is not the case here.)

@oscarbenjamin Is suggesting that this function should accept list of integers as an argument and not an expression because none of the other function in this module accept expression as input. I think this function should be similar and accept list of integers as input.

Reply:
A higher level method might be `expr.congruent(m)` which determines what to do with `expr` to give an answer. In the case of a power, is there any smart thing that can be done with the coefficients of the base so the power wouldn't need to be expanded, e.g. `((x + 1)**20).congruent(13)`?

Reply:
yes, in case of power we can solve for `y**n = a mod m`. where y is the base of the expression.
but for this, the input to the function should be `expr`. And i am not sure hot to implement `expr.congruent(m)` in a function. could you please clarify more on this.

Reply:
This is where design considerations come into play. Although it might only make sense to implement this for Poly, @oscarbenjamin had the right sentiment to get this expressed in terms of coefficients for the function-space for which you were adding the calculation. But I also know that Poly is always going to expand the input and maybe there would be cases where you wouldn't want to expand the input. What I don't know is if this really is going to lead to some simplification in output. I think for now that what you have is fine. The next step might be to add a `congruent` method to `Expr` which (if Expr is a polynomial (univariate?)) or a power of the same (with an integer exponent) would pass the proper coefficients to the function you wrote.

I would prefer that someone else signs off on this PR.

Reply:
Okay, i understand. I will work on congruent method to Expr on a separate pr.
@asmeurer Could you please take a review here?

Reply:
I think `Expr` already has too many methods and a congruent method wouldn't make sense in all cases e.g. `ImmutableMatrix([...]).congruent(Poly(...))`. I think it would be better for congruent to be a function rather than a method.

Reply:
Good point. And I just remembered that Poly has the option to not expand its arg:

```python
>>> Poly((x+1)**20,expand=0)
Poly((x + 1)**20, x + 1, domain='ZZ')
```

So it would be ok to have this as a method on Poly. Though just a general function, like you said, that checks for special cases would work, too. Again: this doesn't need to be part of this PR.